### PR TITLE
Encrypt Node_id and Tor_id at rest

### PR DESCRIPTION
**Pull request by SWvheerden**
## Description
This PR adds in the NodeID and TorID to the wallet's database as encrypted fields 

## Motivation and Context
This allows us to protect the sensitive private keys for the Tari Comms and Tor comms as they are not stored in a plain text file at rest anymore. 

## Types of changes

* [ ] Bug fix (non-breaking change which fixes an issue)
* [x] New feature (non-breaking change which adds functionality)
* [ ] Breaking change (fix or feature that would cause existing functionality to change)
* [ ] Feature refactor (No new feature or functional changes, but performance or technical debt improvements)
* [ ] New Tests
* [ ] Documentation

## Checklist:

* [ ] I'm merging against the `development` branch.
* [ ] I ran `cargo-fmt --all` before pushing.
* [ ] I ran `cargo test` successfully before submitting my PR.
* [ ] I have squashed my commits into a single commit.
* [ ] My change requires a change to the documentation.
* [ ] I have updated the documentation accordingly.
* [ ] I have added tests to cover my changes.

### CODE REVIEW

File: applications/tari_console_wallet/src/init/mod.rs [hunk 1]
```diff
@@ -206,21 +222,83 @@ pub async fn init_wallet(
...
+            },
+            Err(e) => {
+                return Err(ExitCodes::WalletError(format!(
+                    "Error creating Wallet database backends. {}",
+                    e
+                )));
+            },
+        }
+    };
+    debug!(target: LOG_TARGET, "after testnet case");
```
Comment (stringhandler) [inline applications/tari_console_wallet/src/init/mod.rs]: Debugging code?

File: applications/tari_console_wallet/src/init/mod.rs [hunk 2]
```diff
@@ -206,21 +222,83 @@ pub async fn init_wallet(
...
     };
-
     let (wallet_backend, transaction_backend, output_manager_backend, contacts_backend) = backends;
 
     debug!(
         target: LOG_TARGET,
         "Databases Initialized. Wallet encrypted? {}.", wallet_encrypted
-    );
-
+    ); // TODO remove after next TestNet
```
Comment (stringhandler) [inline applications/tari_console_wallet/src/init/mod.rs]: Perhaps some more detail here?

File: applications/tari_console_wallet/src/init/mod.rs [hunk 3]
```diff
@@ -206,21 +222,83 @@ pub async fn init_wallet(
...
+            Err(e) => {
+                return Err(ExitCodes::WalletError(format!(
+                    "Error creating Wallet database backends. {}",
+                    e
+                )));
+            },
+        }
+    };
+    debug!(target: LOG_TARGET, "after testnet case");
+    let _ = wallet_backend.fetch(&DbKey::Identity);
```
Comment (stringhandler) [inline applications/tari_console_wallet/src/init/mod.rs]: Duplicate code

File: applications/tari_console_wallet/src/init/mod.rs [hunk 4]
```diff
@@ -206,21 +222,83 @@ pub async fn init_wallet(
...
-
     let (wallet_backend, transaction_backend, output_manager_backend, contacts_backend) = backends;
 
     debug!(
         target: LOG_TARGET,
         "Databases Initialized. Wallet encrypted? {}.", wallet_encrypted
-    );
-
+    ); // TODO remove after next TestNet
+    if node_identity.is_some() {
```
Comment (stringhandler) [inline applications/tari_console_wallet/src/init/mod.rs]: clippy usually picks this up as `if let Some(x) = ... {`

File: base_layer/wallet/src/storage/database.rs [hunk 5]
```diff
@@ -47,12 +52,20 @@ pub trait WalletBackend: Send + Sync + Clone {
...
     ClientKey(String),
 }
 
 pub enum DbValue {
     CommsSecretKey(CommsSecretKey),
     CommsPublicKey(CommsPublicKey),
+    CommsAddress(Multiaddr),
+    CommsFeatures(u64),
+    Identity(NodeIdentity),
+    TorID(TorIdentity),
```
Comment (stringhandler) [inline base_layer/wallet/src/storage/database.rs]:
```suggestion
    TorId(TorIdentity),
```

File: base_layer/wallet/src/storage/sqlite_db.rs [hunk 6]
```diff
@@ -204,31 +230,93 @@ impl WalletSqliteDatabase {
...
+                WalletSettingSql::new(DbKey::TorID.to_string(), ciphertext_integral_nonce.to_hex()).set(&conn)?;
+            },
+        }
+
+        Ok(())
+    }
+
+    fn get_tor_id(&self, conn: &SqliteConnection) -> Result<Option<DbValue>, WalletStorageError> {
+        let cipher = acquire_read_lock!(self.cipher);
+        if let Some(key_str) = WalletSettingSql::get(format!("{}", DbKey::TorID), &conn)? {
```
Comment (stringhandler) [inline base_layer/wallet/src/storage/sqlite_db.rs]: nit: `.to_string()`

File: base_layer/wallet/src/storage/sqlite_db.rs [hunk 7]
```diff
@@ -344,6 +485,21 @@ impl WalletBackend for WalletSqliteDatabase {
             ckv.set(&conn)?;
         }
 
+        // remove tor id encryption if present
+        let key_str = WalletSettingSql::get(format!("{}", DbKey::TorID), &conn)?;
+        match key_str {
+            Some(v) => {
+                let decrypted_key_bytes = decrypt_bytes_integral_nonce(&cipher, from_hex(v.as_str())?)
+                    .map_err(|_| WalletStorageError::AeadError("Decryption Error".to_string()))?;
```
Comment (stringhandler) [inline base_layer/wallet/src/storage/sqlite_db.rs]: Swallowing an error here

**Review by stringhandler — COMMENTED** [review]
Happy in general, but a couple of clippy warnings and style nits. There are also a number of places where you have swallowed errors

File: base_layer/wallet/src/storage/sqlite_db.rs (hunk 7 above)
Comment (SWvheerden) [reply base_layer/wallet/src/storage/sqlite_db.rs]: Changed the error type here to not swallow any errors anymore

File: base_layer/wallet/src/storage/sqlite_db.rs (hunk 6 above)
Comment (SWvheerden) [reply base_layer/wallet/src/storage/sqlite_db.rs]: Removed all format!("{}" and replaced with to_string()

File: applications/tari_console_wallet/src/init/mod.rs [hunk 8]
```diff
@@ -206,21 +219,84 @@ pub async fn init_wallet(
...
     debug!(
         target: LOG_TARGET,
         "Databases Initialized. Wallet encrypted? {}.", wallet_encrypted
     );
-
+    // TODO remove after next TestNet
+    // If we know node_identity is not passed in anymore we dont have to check if we need to write it to db. This
+    // assumes that if its passed in we need to still save it the database.
+    if node_identity.is_some() {
+        match Arc::try_unwrap(node_identity.unwrap()) {
```
Comment (philipr-za) [inline applications/tari_console_wallet/src/init/mod.rs]: Pretty sure you can do away with all of this unwrapping stuff and just go:
```Rust
(*Arc::make_mut(&mut node_identity)).clone()
```
Assuming you use the suggested `if let Some(x)` above to get inside the option

File: base_layer/wallet/src/storage/sqlite_db.rs [hunk 9]
```diff
@@ -204,33 +229,99 @@ impl WalletSqliteDatabase {
...
+        if let Some(key_str) = WalletSettingSql::get(DbKey::TorId.to_string(), &conn)? {
+            let tor_json = match cipher.as_ref() {
+                None => key_str,
+                Some(cipher) => {
+                    let decrypted_key_bytes = decrypt_bytes_integral_nonce(&cipher, from_hex(&key_str)?)
+                        .map_err(|_| WalletStorageError::AeadError("Decryption Error".to_string()))?;
+                    let tor_json = std::str::from_utf8(&decrypted_key_bytes)
+                        .map_err(|_| WalletStorageError::AeadError("Decryption Error".to_string()))?
+                        .to_string();
+                    // Encryption and decryption breaks the json string, so we fix it.
```
Comment (philipr-za) [inline base_layer/wallet/src/storage/sqlite_db.rs]: This feels weird, the encryption should return the bytes stream byte for byte?

File: applications/tari_console_wallet/src/init/mod.rs (hunk 4 above)
Comment (SWvheerden) [reply applications/tari_console_wallet/src/init/mod.rs]: Did all clippy comments

File: base_layer/wallet/src/storage/sqlite_db.rs (hunk 9 above)
Comment (SWvheerden) [reply base_layer/wallet/src/storage/sqlite_db.rs]: Its because we are encoding a string
so this:
"{\"private_key\":{\"Ed25519V3\":\"aFn0Bod+Vvq8/obUEbVPVsa9Vdk9p44FOavDKzJVoU72dyJ65k9Zy6+NuWrwI1uI51ApzG5IjDUH2NjW9fKmoA==\"},\"service_id\":\"pqdggzs7wlxguwdyynojpez3a7tfzo2zacks2ppivyp4pnsu7lmkpzqd\",\"onion_port\":18101}"
Becomes:
"\"{\\"private_key\\":{\\"Ed25519V3\\":\\"aFn0Bod+Vvq8/obUEbVPVsa9Vdk9p44FOavDKzJVoU72dyJ65k9Zy6+NuWrwI1uI51ApzG5IjDUH2NjW9fKmoA==\\"},\\"service_id\\":\\"pqdggzs7wlxguwdyynojpez3a7tfzo2zacks2ppivyp4pnsu7lmkpzqd\\",\\"onion_port\\":18101}\""

File: applications/tari_console_wallet/src/init/mod.rs (hunk 8 above)
Comment (SWvheerden) [reply applications/tari_console_wallet/src/init/mod.rs]: Clippy was happy with this one,  but I updated the code with (*Arc::make_mut(&mut node_identity)).clone()

File: base_layer/wallet/src/storage/sqlite_db.rs (hunk 9 above)
Comment (stringhandler) [reply base_layer/wallet/src/storage/sqlite_db.rs]: Why not serialize it as bytes with serde?

File: base_layer/wallet/src/storage/sqlite_db.rs [hunk 10]
```diff
@@ -204,33 +229,99 @@ impl WalletSqliteDatabase {
...
     }
+
+    fn get_comms_secret_key(&self, conn: &SqliteConnection) -> Result<Option<DbValue>, WalletStorageError> {
+        let cipher = acquire_read_lock!(self.cipher);
+        if let Some(key_str) = WalletSettingSql::get(DbKey::CommsSecretKey.to_string(), &conn)? {
+            let secret_key = match cipher.as_ref() {
+                None => CommsSecretKey::from_hex(key_str.as_str())?,
+                Some(cipher) => {
+                    let decrypted_key_bytes = decrypt_bytes_integral_nonce(&cipher, from_hex(key_str.as_str())?)
+                        .map_err(|_| WalletStorageError::AeadError("Decryption Error".to_string()))?;
```
Comment (stringhandler) [inline base_layer/wallet/src/storage/sqlite_db.rs]: Still swallowing this error

File: base_layer/wallet/src/storage/sqlite_db.rs [hunk 11]
```diff
@@ -243,6 +334,27 @@ impl WalletBackend for WalletSqliteDatabase {
                     Some(DbValue::ClientValue(v.value))
                 },
             },
+            DbKey::CommsAddress => self.get_comms_address(&conn)?,
+            DbKey::Identity => {
+                let secret_key = match self.get_comms_secret_key(&conn)? {
+                    Some(DbValue::CommsSecretKey(v)) => v,
+                    _ => return Ok(None),
+                };
+                let address = match self.get_comms_address(&conn)? {
```
Comment (stringhandler) [inline base_layer/wallet/src/storage/sqlite_db.rs:338]: Why not rather return a valid `Option<CommsAddress>` here instead of wrapping it in a `DbValue`? Seems to add a lot of code in the checking stage

File: base_layer/wallet/src/storage/sqlite_db.rs [hunk 12]
```diff
@@ -187,6 +192,26 @@ impl WalletSqliteDatabase {
         Ok(())
     }
 
+    fn set_node_id(&self, node_id: &NodeIdentity, conn: &SqliteConnection) -> Result<(), WalletStorageError> {
+        let cipher = acquire_read_lock!(self.cipher);
+        WalletSettingSql::new(DbKey::CommsPublicKey.to_string(), node_id.public_key().to_hex()).set(&conn)?;
+        WalletSettingSql::new(DbKey::CommsAddress.to_string(), node_id.public_address().to_string()).set(&conn)?;
+        WalletSettingSql::new(DbKey::CommsFeatures.to_string(), node_id.features().bits().to_string()).set(&conn)?;
```
Comment (stringhandler) [inline base_layer/wallet/src/storage/sqlite_db.rs:199]: Why are features saved to the DB? Are these configurable?

File: base_layer/wallet_ffi/src/lib.rs [hunk 13]
```diff
@@ -2339,7 +2340,7 @@ pub unsafe extern "C" fn transport_tcp_create(
 pub unsafe extern "C" fn transport_tor_create(
     control_server_address: *const c_char,
     tor_cookie: *const ByteVector,
-    tor_identity: *const ByteVector,
+    _tor_identity: *const ByteVector,
```
Comment (stringhandler) [inline base_layer/wallet_ffi/src/lib.rs]: Probably best to just remove it

File: base_layer/wallet/src/storage/sqlite_db.rs (hunk 11 above)
Comment (philipr-za) [reply base_layer/wallet/src/storage/sqlite_db.rs:338]: This is part of this modular storage backend pattern that we conceived in the very beginning of the project to allow for multiple backends to be written that implement the `fetch(...)` and `write(...)` methods of the trait. But I agree I think the boilerplate this creates is not worth the benefit for a potential easier future integration of some other backend. Though removing it now is a cleanup refactor in itself so 🤷‍♂️

Comment (stringhandler) [reply base_layer/wallet/src/storage/sqlite_db.rs:338]: Fair, but, I was more referring to the fact that `get_comms_address` could be written to return the result in the correct type (i.e. `CommsAddress`), and only the instance that needed to return a `DbValue` (i.e. `fetch`) wrap it.

Comment (SWvheerden) [reply base_layer/wallet/src/storage/sqlite_db.rs:338]: This is the same schematics we use all over the code base, from the BaseNode to the wallet for db write transactions.  I only did it this way to be consistent with the rest of the code.

File: base_layer/wallet/src/storage/sqlite_db.rs (hunk 10 above)
Comment (SWvheerden) [reply base_layer/wallet/src/storage/sqlite_db.rs]: I removed these all as well.

File: base_layer/wallet_ffi/src/lib.rs (hunk 13 above)
Comment (philipr-za) [reply base_layer/wallet_ffi/src/lib.rs]: careful, this will break the Swift and JNI wrappers if we don't update those at the same time.

File: base_layer/wallet/src/storage/sqlite_db.rs (hunk 12 above)
Comment (SWvheerden) [reply base_layer/wallet/src/storage/sqlite_db.rs:199]: features is, 
The other two is purely convenance so that we dont have to recalc them. I added everything from the node_id.json file.

File: base_layer/wallet_ffi/src/lib.rs [hunk 14]
```diff
@@ -2951,6 +2970,15 @@ pub unsafe extern "C" fn wallet_create(
 
             match w {
                 Ok(mut w) => {
+                    // lets ensure the wallet tor_id is saved
+                    if let Some(hs) = w.comms.hidden_service() {
+                        if let Err(e) = runtime.block_on(w.db.set_tor_identity(hs.tor_identity().clone())) {
+                            warn!(
+                                target: LOG_TARGET,
+                                "Could not restart transaction negotiation protocols: {}", e
```
Comment (philipr-za) [inline base_layer/wallet_ffi/src/lib.rs]: This message doesn't seem related to the error.

File: base_layer/wallet/src/storage/sqlite_db.rs (hunk 11 above)
Comment (SWvheerden) [reply base_layer/wallet/src/storage/sqlite_db.rs:338]: `fn get_comms_address` is a private function. 
You either have to unwrap the option here, and use above straight
OR
Use here straight, and wrap above

Comment (stringhandler) [reply base_layer/wallet/src/storage/sqlite_db.rs:338]: I disagree that it is like that in the code base, and lines 330 and 324 are both examples of doing it correctly, and `get_comms_address` is a new method created in this PR. I would say change line 337 to 
```
DbKey::CommsAddress => self.get_comms_address(&conn)?.map(|v| DbValue::CommsAddress(v)),
```
and `get_comms_address` to:
```
fn get_comms_address(&self, conn: &SqliteConnection) -> Result<Option<CommsAddress>, WalletStorageError> 
```

File: base_layer/wallet/src/storage/sqlite_db.rs (hunk 9 above)
Comment (SWvheerden) [reply base_layer/wallet/src/storage/sqlite_db.rs]: I will try and do that when saving as encrypted
